app/frontend/components/shared/revisions/revision-modal.tsx

```diff
@@ -0,0 +1,211 @@
+import {
+  Box,
+  Button,
+  Divider,
+  Flex,
+  FormControl,
+  FormHelperText,
+  FormLabel,
+  Heading,
+  Modal,
+  ModalBody,
+  ModalCloseButton,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+  Select,
+  Spacer,
+  Textarea,
+  useDisclosure,
+} from "@chakra-ui/react"
+import { Trash } from "@phosphor-icons/react"
+import React, { useState } from "react"
+import { UseFieldArrayReturn } from "react-hook-form"
+import { useTranslation } from "react-i18next"
+import { useMst } from "../../../setup/root"
+import { EReasonCode } from "../../../types/enums"
+import { IFormIORequirement, IRevisionRequest } from "../../../types/types"
+import { singleRequirementFormJson } from "../../../utils/formio-helpers"
+import { IRevisionRequestForm } from "../../domains/permit-application/revision-sidebar"
+import { Form } from "../chefs"
+
+export interface IRevisionModalProps extends Partial<ReturnType<typeof useDisclosure>> {
+  requirementJson: IFormIORequirement
+  submissionJson: any
+  revisionRequest: IRevisionRequest
+  useFieldArrayMethods: UseFieldArrayReturn<IRevisionRequestForm, "revisionRequestsAttributes", "fieldId">
+  onSave: () => Promise<void>
+  isRevisionsRequested?: boolean
+  forSubmitter?: boolean
+}
+
+export const RevisionModal: React.FC<IRevisionModalProps> = ({
+  requirementJson,
+  submissionJson,
+  isOpen,
+  onOpen,
+  onClose,
+  revisionRequest,
+  useFieldArrayMethods,
+  onSave,
+  isRevisionsRequested,
+  forSubmitter,
+}) => {
+  const { t } = useTranslation()
+  const [reasonCode, setReasonCode] = useState<EReasonCode | "">(revisionRequest?.reasonCode ?? "")
+  const [comment, setComment] = useState<string>(revisionRequest?.comment ?? "")
+
+  const { update, append, fields } = useFieldArrayMethods
+
+  const { userStore } = useMst()
+  const { currentUser } = userStore
+
+  const className = `formio-component-${requirementJson.key}`
+  const elements = document.getElementsByClassName(className)
+
+  const resetFields = () => {
+    setReasonCode("")
+    setComment("")
+  }
+
+  const handleClose = () => {
+    resetFields()
+    onClose()
+  }
+  const index = fields.findIndex((field) => field.id === revisionRequest?.id)
+
+  const handleUpsert = () => {
+    if (reasonCode && requirementJson) {
+      const newItem = {
+        id: revisionRequest?.id,
+        userId: currentUser.id,
+        reasonCode,
+        requirementJson,
+        submissionJson,
+        comment,
+      }
+      if (revisionRequest) {
+        // Item exists, replace it
+        update(index, newItem)
+      } else {
+        // Item does not exist, append it
+        append(newItem)
+      }
+
+      onSave().then(() => {
+        elements?.[0]?.classList?.add("revision-requested")
+        handleClose()
+      })
+    }
```

Comment:
instead of having nested conditionals. It is better to early return when possible for better readability